storage: make replica.GetReplica()s interface more idiot-proof

andreimatei opened the pull request:
CC @BramGruneir 

I don't really know what I'm doing. Is the check in `redirectOnOrAcquireLease` a reasonable place to put it?
Also it's unclear to me what goes on / whether anything should be done on the sender side - after a raft request with `REMOVE_REPLICA` is proposed, should the node that's proposing that command not use that replica any more even before the command applies, if it's local?

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7148)

<!-- Reviewable:end -->

andreimatei committed:
storage: make replica.GetReplica()s interface more idiot-proof

Force people to handle the case where it used to return nil - the
replica has been removed from the store.

tamird commented:
first commit is a minor improvement, but that interface is still a mess. I'm inclined to advise that we not touch it in this PR but file a new contributor issue to clean it up.

can you expand on what it is you're fixing in the second commit message? a test would of course be best, but even without a test it'd be nice to be able to tell what the impetus for this PR is.

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> #### storage: fix trying to clean up intents with a replica that was just removed
> 
> CC @BramGruneir 
> 
> 
> 
> I don't really know what I'm doing. Is the check in `redirectOnOrAcquireLease` a reasonable place to put it?
> 
> Also it's unclear to me what goes on / whether anything should be done on the sender side - after a raft request with `REMOVE_REPLICA` is proposed, should the node that's proposing that command not use that replica any more even before the command applies, if it's local?
> 
> 
> 
</details>


---

Reviewed 5 of 5 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 739 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFWee9bvtT6_9j_DU:-KJrFWefQjC-dwXP-dMW:678108080) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/replica.go#L739)):*
> ```Go
> 
> func (e *errReplicaNotInRaftGroup) Error() string {
> 	return fmt.Sprintf("Replica in store %s not found in RangeDescriptor %+v."+
> ```

s/%s/%d/

---

*[storage/replica.go, line 740 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFYc_p-ygmZEhgVYJ:-KJrFYc_p-ygmZEhgVYK:1256602193) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/replica.go#L740)):*
> ```Go
> func (e *errReplicaNotInRaftGroup) Error() string {
> 	return fmt.Sprintf("Replica in store %s not found in RangeDescriptor %+v."+
> 		"Maybe it's been removed from the Raft group already.", e.StoreID, e.RangeDesc)
> ```

is this speculation helpful?

---

*[storage/stores.go, line 211 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFyZGOcjhIGeMKwpk:-KJrFyZGOcjhIGeMKwpl:-2026831384) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/stores.go#L211)):*
> ```Go
> 			if err != nil {
> 				if _, ok := err.(*errReplicaNotInRaftGroup); ok {
> 					// Apparently rng.GetReplica() can return an error, even though
> ```

why is this comment helpful? LookupReplica doesn't check that the replica has an entry in the range descriptor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-:-KJrGSLt0eWRCOE5yjKa:1111374265)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
The first commit looks like an improvement (though I personally think a sentinel error would be enough, but what gives).

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> first commit is a minor improvement, but that interface is still a mess. I'm inclined to advise that we not touch it in this PR but file a new contributor issue to clean it up.
> 
> can you expand on what it is you're fixing in the second commit message? a test would of course be best, but even without a test it'd be nice to be able to tell what the impetus for this PR is.
</details>


---

Reviewed 5 of 5 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 633 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJuwCQ-nw-cTXiriwgI:-KJuwCQ-nw-cTXiriwgJ:1665766342) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bd129d86364a38de0e8cfd4ea97d1bb2e76be5d/storage/replica.go#L633)):*
> ```Go
> 				return roachpb.NewError(r.newNotLeaderError(lease, r.store.StoreID()))
> 			}
> 			// Check that this replica is still a member of the Raft group. If it's
> ```

If anything, this check should also sit squarely in the path of requesting the leader lease (you're in the case in which we already have it).
As Tamir suggested above, I don't think we should start sprinkling random checks all over the place. Various goroutines might be in the middle of using the replica when it is removed, so we should do something more general.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-:-KJuwqGRHqiXg-VQyXry:-1369661005)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
ok, removed the commit adding the random "am I still a good replica" checks with a more targeted new commit that prevents a particular case of using a remove replica: processing the intents after the replica removal.

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The first commit looks like an improvement (though I personally think a sentinel error would be enough, but what gives).
</details>


---

Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 633 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJuwCQ-nw-cTXiriwgI:-KJuxhomE7wJe8lJbDQk:2085334901) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bd129d86364a38de0e8cfd4ea97d1bb2e76be5d/storage/replica.go#L633)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> If anything, this check should also sit squarely in the path of requesting the leader lease (you're in the case in which we already have it).
> As Tamir suggested above, I don't think we should start sprinkling random checks all over the place. Various goroutines might be in the middle of using the replica when it is removed, so we should do something more general.
</details>
So what do you suggest? Moving the check inside `getLeaderLease()`? That would put the check in the path both inside and outside of Raft. Or not having the check at all?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-:-KJwUzAwmFCXhAypzeu-:1415174821)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/replica_command.go, line 525 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KK081sgHkNDc14sDJGk:-KK081sgHkNDc14sDJGl:-229063531) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c54a22b07f1ca37395fda94e048551b6205cf6f/storage/replica_command.go#L525)):*
> ```Go
> 		if crt := args.InternalCommitTrigger.GetChangeReplicasTrigger(); crt != nil {
> 			if crt.ChangeType == roachpb.REMOVE_REPLICA {
> 				externalIntents = externalIntents[:0]
> ```

nit: not nil?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 633 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJuwCQ-nw-cTXiriwgI:-KK0QEyh14GPW-zyxhCb:-1659572968) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bd129d86364a38de0e8cfd4ea97d1bb2e76be5d/storage/replica.go#L633)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> So what do you suggest? Moving the check inside `getLeaderLease()`? That would put the check in the path both inside and outside of Raft. Or not having the check at all?
</details>
Not having it here at all, since it's a patchwork solution.

---

*[storage/replica_command.go, line 520 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KK0Oj5KzBBPw5NQhUFI:-KK0Oj5KzBBPw5NQhUFJ:-1796858631) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c54a22b07f1ca37395fda94e048551b6205cf6f/storage/replica_command.go#L520)):*
> ```Go
> 			return reply, nil, err
> 		}
> 		// If we've just remove the replica, don't return any intents from this
> ```

s/remove/removed/

---

*[storage/replica_command.go, line 525 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KK081sgHkNDc14sDJGk:-KK0P3MVS6Bf8LG4ssyj:698050350) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c54a22b07f1ca37395fda94e048551b6205cf6f/storage/replica_command.go#L525)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: not nil?
</details>
Also expected `nil`. I know I brought this change up, but now that I think about it again it's not that great - you're worried about the case in which the leader removes itself, but this change doesn't resolve the external intents at all, regardless of which scenario. Hence, this is net-negative.

I just filed #7169 - I think the best way to unblock you is to remove this part of the code, bring back that check in the `processIntentsAsync` (which checks for the replica descriptor), mentioning a TODO pointing to the issue I created.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
removed the last commit. The only thing remaining in this PR is changing the `GetReplica` interface

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> ok, removed the commit adding the random "am I still a good replica" checks with a more targeted new commit that prevents a particular case of using a remove replica: processing the intents after the replica removal.
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 739 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFWee9bvtT6_9j_DU:-KKAX1YKWi-1xsMsX4Gn:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/replica.go#L739)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/%s/%d/
</details>
Done.

---

*[storage/replica.go, line 740 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFYc_p-ygmZEhgVYJ:-KKAX2uQp_sjGzxGiQGI:1309408349) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/replica.go#L740)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this speculation helpful?
</details>
I think so... It would've helped me figure out what's going on.

---

*[storage/stores.go, line 211 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFyZGOcjhIGeMKwpk:-KKAXqWQVpxgEdZap0si:506201268) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/stores.go#L211)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is this comment helpful? LookupReplica doesn't check that the replica has an entry in the range descriptor.
</details>
I've removed the comment. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-:-KKAbgM-NwfaQaf4eZOe:-2044048872)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> removed the last commit. The only thing remaining in this PR is changing the `GetReplica` interface
</details>


---

Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-:-KKAdEyfJ00rKssPxtNw:1417088721)*
<!-- Sent from Reviewable.io -->

tamird commented:
PR description is stale (so the merge commit will be misleading)

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> :lgtm:
</details>


---

Reviewed 1 of 6 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 740 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFYc_p-ygmZEhgVYJ:-KKAcyRbubFHHWMqoGbZ:172430290) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/replica.go#L740)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think so... It would've helped me figure out what's going on.
</details>
Will it help the next person who encounters this error for some other reason?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-:-KKAdaN3MwHmGkynDgox:1098369525)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 6 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
changed the description. I couldn't figure out how to do it before from Reviewable directly. Can you?

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> PR description is stale (so the merge commit will be misleading)
</details>


---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 740 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFYc_p-ygmZEhgVYJ:-KKAeUa8pRYybVRet3tS:1711027228) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/replica.go#L740)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Will it help the next person who encounters this error for some other reason?
</details>
yes :)
The likely cause of seeing this error is a race with removing a replica, which this msg suggests. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-:-KKAeJVG1gN-3MN0pFcf:-1648757591)*
<!-- Sent from Reviewable.io -->

tamird commented:
reviewable doesn't mutate github state, so it seems expected that you can't change that from here.

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> changed the description. I couldn't figure out how to do it before from Reviewable directly. Can you?
</details>


---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 740 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFYc_p-ygmZEhgVYJ:-KKAmGRIjTrLRUianlsc:1198263755) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/replica.go#L740)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> yes :)
> The likely cause of seeing this error is a race with removing a replica, which this msg suggests. 
</details>
extremely -1 on this. you're asking for rot here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-:-KKAm9Y4GQDTLq7Wl4ZW:-130646023)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 740 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFYc_p-ygmZEhgVYJ:-KKAtpD0mLwsjog5ZKRI:2038268067) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/replica.go#L740)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> extremely -1 on this. you're asking for rot here.
</details>
this error is returned when you're assuming a something (i.e. a replica) is in a set (i.e. a range descriptor) and in fact it's not. The message is suggesting that maybe it was removed. The other possibility is that it was never added. I chose to hint at one of these possibilities in particular. I don't really see what could rot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 740 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFYc_p-ygmZEhgVYJ:-KKC8R5otWKZrzk7mcK1:1075948605) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/replica.go#L740)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> this error is returned when you're assuming a something (i.e. a replica) is in a set (i.e. a range descriptor) and in fact it's not. The message is suggesting that maybe it was removed. The other possibility is that it was never added. I chose to hint at one of these possibilities in particular. I don't really see what could rot.
</details>
I tend to agree with Tamir here, but feel less strongly. Prior to preemptive snapshots, the only time you'd see this was when the range was already removed; you never actually had a Replica object for a range you were not yet a member of. But now you can, and there's a good chance that the next time we see this error it will be "not yet added" instead of "already removed". 

---

*[storage/replica.go, line 760 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKC9QhIgtl0x0K2l73r:-KKC9QhIgtl0x0K2l73s:-1854847733) ([raw file](https://github.com/cockroachdb/cockroach/blob/4392b8d9f1a148d6f297fdecaac7349c355825c2/storage/replica.go#L760)):*
> ```Go
> }
> 
> type errReplicaNotInRaftGroup struct {
> ```

You don't actually look at the raft group. `errReplicaNotInRange`? Or just `errReplicaNotFound`, although that may be confusingly similar to `roachpb.RangeNotFoundError`, especially since `Store.GetReplica` returns a `RangeNotFoundError` instead of anything that mentions replicas. 

---

*[storage/replica.go, line 771 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKC84uw6hrwH4O_U0YK:-KKC84ux_VHeZ2C60AeO:-1871378275) ([raw file](https://github.com/cockroachdb/cockroach/blob/4392b8d9f1a148d6f297fdecaac7349c355825c2/storage/replica.go#L771)):*
> ```Go
> 
> // GetReplica returns the replica for this range from the range descriptor.
> // Returns an errReplicaNotInRaftGroup if the replica is not found.
> ```

Add a star before `errReplicaNotInRaftGroup` - it's a pointer, and this matters for type casts. 

If you document this as "Returns nil, *errReplicaNotInRaftGroup if the replica is not found", then you could get rid of the `replica = nil` line in `stores.go`. 

---

*[storage/store.go, line 2381 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKC7OpaUN40XzNyTjcg:-KKC7OpaUN40XzNyTjch:2025564249) ([raw file](https://github.com/cockroachdb/cockroach/blob/4392b8d9f1a148d6f297fdecaac7349c355825c2/storage/store.go#L2381)):*
> ```Go
> 	newStoreRangeSet(s).Visit(func(r *Replica) bool {
> 		desc, err := r.GetReplica()
> 		if _, replicaGone := err.(*errReplicaNotInRaftGroup); !r.IsInitialized() || replicaGone {
> ```

The way that `r.IsInitialized` appears in the middle of the `replicaGone` check is weird (it was weird before, but it's worse now). Maybe add a separate `if r.IsInitialized() { return true}` before going into `r.GetReplica()`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-:-KKCA3KVeguI9U-pW1E8:1417088721)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 740 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KJrFYc_p-ygmZEhgVYJ:-KKFE_3lS0sTwToIFuC3:467769323) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea36f72805978cee1f8bdf0cc1694bec2d43c07b/storage/replica.go#L740)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I tend to agree with Tamir here, but feel less strongly. Prior to preemptive snapshots, the only time you'd see this was when the range was already removed; you never actually had a Replica object for a range you were not yet a member of. But now you can, and there's a good chance that the next time we see this error it will be "not yet added" instead of "already removed". 
</details>
ok, removed

---

*[storage/replica.go, line 760 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKC9QhIgtl0x0K2l73r:-KKFEOeSw3kIs_a3cnkt:-2136103039) ([raw file](https://github.com/cockroachdb/cockroach/blob/4392b8d9f1a148d6f297fdecaac7349c355825c2/storage/replica.go#L760)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> You don't actually look at the raft group. `errReplicaNotInRange`? Or just `errReplicaNotFound`, although that may be confusingly similar to `roachpb.RangeNotFoundError`, especially since `Store.GetReplica` returns a `RangeNotFoundError` instead of anything that mentions replicas. 
</details>
done. `errReplicaNotInRange`

---

*[storage/replica.go, line 771 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKC84uw6hrwH4O_U0YK:-KKFFUO0jT0KT284uTqI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4392b8d9f1a148d6f297fdecaac7349c355825c2/storage/replica.go#L771)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Add a star before `errReplicaNotInRaftGroup` - it's a pointer, and this matters for type casts. 
> 
> If you document this as "Returns nil, *errReplicaNotInRaftGroup if the replica is not found", then you could get rid of the `replica = nil` line in `stores.go`. 
</details>
Done.

---

*[storage/store.go, line 2381 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKC7OpaUN40XzNyTjcg:-KKFGFGGrO6g7MDCLglC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4392b8d9f1a148d6f297fdecaac7349c355825c2/storage/store.go#L2381)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The way that `r.IsInitialized` appears in the middle of the `replicaGone` check is weird (it was weird before, but it's worse now). Maybe add a separate `if r.IsInitialized() { return true}` before going into `r.GetReplica()`. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2381 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKC7OpaUN40XzNyTjcg:-KKFNnV4uxzmKFEvY8xw:2078613130) ([raw file](https://github.com/cockroachdb/cockroach/blob/4392b8d9f1a148d6f297fdecaac7349c355825c2/storage/store.go#L2381)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Done.
</details>
what if we get a different kind of error?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2400 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKFiAMbYxrcLWib70ql:-KKFiAMcCJ3-DBkyL2UU:400709739) ([raw file](https://github.com/cockroachdb/cockroach/blob/37aea883ad238ac04fbd4033cdcbd8afc22d425c/storage/store.go#L2400)):*
> ```Go
> 		}
> 		desc, err := r.GetReplica()
> 		if _, ok := err.(*errReplicaNotInRange); ok {
> ```

nest this under the nit check like you do in storage/stores.go 

---

*[storage/store.go, line 2404 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKFhzYvv-6OgW7cNzJt:-KKFhzYvv-6OgW7cNzJu:-842975055) ([raw file](https://github.com/cockroachdb/cockroach/blob/37aea883ad238ac04fbd4033cdcbd8afc22d425c/storage/store.go#L2404)):*
> ```Go
> 		}
> 		if err != nil {
> 			panic(fmt.Sprintf("unexpected error: %s", err))
> ```

log.Fatalf

---

*[storage/stores.go, line 209 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKFi4dZQummyeL_3HT9:-KKFi4dZQummyeL_3HTA:-1730229459) ([raw file](https://github.com/cockroachdb/cockroach/blob/37aea883ad238ac04fbd4033cdcbd8afc22d425c/storage/stores.go#L209)):*
> ```Go
> 			rangeID = rng.RangeID
> 			replica, err = rng.GetReplica()
> 			if err != nil {
> ```

need to panic here too

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 5 of 6 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2400 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKFiAMbYxrcLWib70ql:-KKFkX5RmXCqzWJAtOSQ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/37aea883ad238ac04fbd4033cdcbd8afc22d425c/storage/store.go#L2400)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nest this under the nit check like you do in storage/stores.go 
</details>
Done.

---

*[storage/store.go, line 2404 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKFhzYvv-6OgW7cNzJt:-KKFkXMieCgz2P7VJqln:-1965035554) ([raw file](https://github.com/cockroachdb/cockroach/blob/37aea883ad238ac04fbd4033cdcbd8afc22d425c/storage/store.go#L2404)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> log.Fatalf
</details>
what's the rule of thumb? panic with errs, but log if you need formatting?

---

*[storage/stores.go, line 209 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKFi4dZQummyeL_3HT9:-KKFkf7ZvGYyynmSRX7x:103289367) ([raw file](https://github.com/cockroachdb/cockroach/blob/37aea883ad238ac04fbd4033cdcbd8afc22d425c/storage/stores.go#L209)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> need to panic here too
</details>
I considered it, but this function has the option to return an error which is how we generally propagate errors-because-of-bugs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2404 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKFhzYvv-6OgW7cNzJt:-KKFlgQyZ_ACrAOtu38U:848333972) ([raw file](https://github.com/cockroachdb/cockroach/blob/37aea883ad238ac04fbd4033cdcbd8afc22d425c/storage/store.go#L2404)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> what's the rule of thumb? panic with errs, but log if you need formatting?
</details>
not sure there's a rule of thumb, but looks odd to panic(fmt.Sprintf(...))

---

*[storage/stores.go, line 209 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148#-KKFi4dZQummyeL_3HT9:-KKFlbPDTT_V8jj70RDK:798790894) ([raw file](https://github.com/cockroachdb/cockroach/blob/37aea883ad238ac04fbd4033cdcbd8afc22d425c/storage/stores.go#L209)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I considered it, but this function has the option to return an error which is how we generally propagate errors-because-of-bugs.
</details>
oh, I see; you return the error if it's not the permitted one. very good.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7148)*
<!-- Sent from Reviewable.io -->
